### PR TITLE
Allow payload to throw exceptions

### LINKED ISSUE #5: Failure to parse body is hard to identify
When the body cannot be parsed the result is an empty HTTP 400 response, with no body. This makes it extremely hard to track where the source of the error is.

### PR DESCRIPTION
**Pull request by shadowhand**
Returning a HTTP 400 response with no body makes it harder to track when
the request body contains an error, and what that error is.

Fixes #5

### CODE REVIEW

**Comment by shadowhand** [comment]
I don't know how to allow for throwing exceptions with CSV, as the method used doesn't appear to throw exceptions no matter what is thrown at it.

**Review by oscarotero — COMMENTED** [review]
As commented in #5, I was thinking in using a more generical HttpErrorException, but like to know other opinions. In case of going for a more specific solution, and due only json throw exceptions, I'd bet for a JsonPayloadException. This allows to extend this package with more formats, or move some formats to different packages easily.

**Comment by shadowhand** [comment]
I think a single `PayloadException` is easier to track and the specific type is available with `$e->getCode()`. Different packages could extend it to provide more types.

**Comment by oscarotero** [comment]
Technically, `$e->getCode()` should returns the value of `json_last_error()`.
Other problem I see is the three formats are coupled to the same exception (The exception has constants to identify the format).
The HttpErrorException prevent that, because you can pass the original exception as the previous exception, so no coupling there. It also could work better in the future, if php 7.3 throws exceptions in json, or if CsvParser switch to [league/csv](https://github.com/thephpleague/csv) that throws exceptions too. The `context` field is used to set extra data in the exceptions, in order to log the errors using a `psr/logger` implementation.

**Comment by shadowhand** [comment]
@oscarotero maybe we should target this for version 2.0.0 and do a number of changes:

- strict types
- use `league/csv` for csv parsing
- use `HttpErrorException`
- throw exceptions by default

**Comment by oscarotero** [comment]
Agree, my only concern is whether include `league/csv` as a dependency. The most used parser is JsonPayload, so most users will install `league/csv` without need it. This is the reason I didn't use it yet.

**Comment by shadowhand** [comment]
I think `league/csv` should be recommendation, not a requirement.

**Comment by shadowhand** [comment]
I think `league/csv` should be recommendation, not a requirement. Or we just break this up into multiple packages: `middlewares/payload-json`, `middlewares/payload-csv` and the base package only deals with URL encoded payload.

**Comment by oscarotero** [comment]
> Or we just break this up into multiple packages:

Yes, that's more or less what I was thinking (I thought create three payload packages and deprecate this, but your suggestion of keeping this for url encoded payload is a better idea). If you want to work on it, that would be great.

**Comment by oscarotero** [comment]
Or, thinking a bit more a bout this, maybe only create `middlewares/payload-csv` because it's the only package that need dependencies and keep json payload here.

**Comment by shadowhand** [comment]
@oscarotero updated to reflect all changes.

File: CHANGELOG.md [hunk 1]
```diff
@@ -5,6 +5,16 @@ All notable changes to this project will be documented in this file.
...
 
+## [2.0.0] - ???
+
+### Added
+
+- Allow parsing to throw exceptions
+
+### Changed
+
+- Moved CSV parsing to a separate package
```
Comment (oscarotero) [inline CHANGELOG.md]: Should we put a link to the new package?

File: README.md [hunk 2]
```diff
@@ -55,11 +55,18 @@ To configure all Content-Type headers allowed in the request. By default is `app
 
 To override the previous parsed body if exists (`false` by default)
 
+#### `throwExceptions($throw = true)`
+
+Allow parsing errors to throw exceptions (`false` by default)
+
+When enabled failure to parse the body will throw a `Middlewares\PayloadException`.
```
Comment (oscarotero) [inline README.md]: The correct exception class is `Middlewares\Utils\PayloadException`

**Review by oscarotero — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 1 above)
Comment (shadowhand) [reply CHANGELOG.md]: It's linked in the readme.

Comment (oscarotero) [reply CHANGELOG.md]: ah, ok

**Comment by shadowhand** [comment]
@oscarotero okay, updated again to add strict types everywhere.

**Review by oscarotero — APPROVED** [review]

**Pull request merged**